engine/apps/grafana_plugin/helpers/client.py

```diff
@@ -355,19 +355,47 @@ def get_instance_info(
         return data
 
     def get_instances(self, query: str, page_size=None):
+        MAX_RETRIES = 3
+
+        missing_next_cursor_key_sentinel = "EMPTYCURSOR"
+        retry_count = 0
+
         if not page_size:
             page, _ = self.api_get(query)
             yield page
         else:
+            previous_cursor = None
             cursor = 0
             while cursor is not None:
                 if query:
-                    page_query = query + f"&cursor={cursor}&pageSize={page_size}"
+                    page_query = f"{query}&cursor={cursor}&pageSize={page_size}"
                 else:
                     page_query = f"?cursor={cursor}&pageSize={page_size}"
-                page, _ = self.api_get(page_query)
-                yield page
-                cursor = page["nextCursor"]
+                page, call_status = self.api_get(page_query)
+
+                previous_cursor = cursor
+                cursor = page.get("nextCursor", missing_next_cursor_key_sentinel)
+
+                if retry_count == MAX_RETRIES:
+                    break
+                elif cursor == missing_next_cursor_key_sentinel:
```

Comment:
`nextCursor` can actually be `None` (this means there's no more pages to paginate), hence why I added a sentinel value for when the `nextCursor` key is missing from the response data